### PR TITLE
Changes to allow editable dataset name

### PR DESCRIPTION
**Pull request by jayasimhankv**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### **Problem statement:** 
We have seen the need for business users to be able to maintain an editable dataset name (aka “business name”), which is separate from the default dataset properties’ name. A previous PR was submitted back in late 2022 which also attempted at adding both editable name for the dataset as well as editable label for schema fields. [Original PR: https://github.com/datahub-project/datahub/pull/6171]

This PR is only confined to being able to add an editable name to a dataset. Please review and let me know if there are any questions/changes.

### High level changes:
**Model**:
- Include name field in EditableDatasetProperties (currently has only description)

**UI/app changes**:
- Change entityHeader.tsx to allow editing dataset’s name based on a privilege check. This provides the ability to edit the name in the header and call the mutation to save it.
- Change datasetEntity.tsx’s displayName property to give preference to editable properties’ name field if one exists.

**UI/graphql changes**:
- Include name field in editableProperties on various fragments, so it can be displayed in the UI.

**Resolver changes**:
- Allow name to be updated for dataset entity.
- Add a new method updateDatasetName that checks for the privilege and then updates the name field in editable dataset properties.
- Mapper changes to get and set name field from and to editableDatasetProperties aspect.

**Working Examples**:
_With the EditProperties privilege granted to the user, the edit icon is shown in entity header, allowing the user to update it._
image

image

_With the editable properties’ name updated in the entity’s aspect, the updated editable name (aka “business name”) is displayed._
image

_Search page shows an example of two datasets - one with the default dataset properties’ name and one with editable dataset properties’s name updated earlier:_
image

_Search example showing look up by editable properties’ name:_
image
 
_Search example showing look up by non-editable properties’ name:_
image

**_Example of a user that does **not** possess the EditProperties privilege - the edit icon in dataset’s header is not provided:_**
image

_Graphql mutation call for a user that has permissions to update the dataset name:_
image

_UI page shows updated name_:
image

_Entity json data - the dataset properties’ name is left intact and only the editable dataset properties’ name is updated as seen below:_
image

**_Graphql mutation call returning unauthorized exception for a user that does NOT have permissions to update the dataset name:_**
image

## Summary by CodeRabbit

- **New Features**
  - Introduced the ability to edit dataset names, configurable through the `editableDatasetNameEnabled` flag.

- **Improvements**
  - Enhanced dataset properties to support editable names.
  - Updated GraphQL queries and fragments to manage editable dataset names effectively.

- **API Updates**
  - Added `name` field to `DatasetEditableProperties` in GraphQL schemas and API responses.

- **Configuration**
  - Integrated `editableDatasetNameEnabled` into application configuration settings.

### CODE REVIEW

File: metadata-service/restli-api/src/main/snapshot/com.linkedin.platform.platform.snapshot.json [hunk 1]
```diff
@@ -843,10 +843,10 @@
       "PRE" : "Designates pre-production fabrics",
       "PROD" : "Designates production fabrics",
       "QA" : "Designates quality assurance fabrics",
+      "RVW" : "Designates review fabrics",
       "STG" : "Designates staging fabrics",
       "TEST" : "Designates testing fabrics",
-      "UAT" : "Designates user acceptance testing fabrics",
-      "RVW" : "Designates review fabrics"
+      "UAT" : "Designates user acceptance testing fabrics"
```
Comment (gabe-lyons) [inline metadata-service/restli-api/src/main/snapshot/com.linkedin.platform.platform.snapshot.json:849]: do you know why these enum values were moved aroud?

**Review by jjoyce0510 — COMMENTED** [review]
Overall this makes a LOT of sense. 

One request I'd have is to add this edit button BEHIND A FEATURE FLAG (can add via application.yaml > featureFlags) so that we can disable this by default, unless people want it enabled.

I think it's somewhat risky from a product pov to allow any dataset names to be editable by default (and perhaps confusing), but if an org wants to opt in this makes a lot a lot of sense.

Step to add a feature flag:

1. Add application.yaml > featureFlags > editableDatasetNames flag
2. Add to FeatureFlags.java
3. Add to app.graphql AppConfig type in GraphQL the new editableDatasetName field
4. Edit AppConfigResolver to map the featureFlag object from YAML to the GraphQL editableDatasetName field.
5. Fetch new field inside of app.graphql on the datahub-web-react side 
6. Add it to DEFAULT_APP_CONFIG in appConfigContext.tsx
7. Finally, inside the EntityHeader.tsx component, add useAppConfig to get the feature flags
8. Show or hide the edit capability based on whether the flag is true

**Comment by jayasimhankv** [comment]
> Overall this makes a LOT of sense.
> 
> One request I'd have is to add this edit button BEHIND A FEATURE FLAG (can add via application.yaml > featureFlags) so that we can disable this by default, unless people want it enabled.
> 
> I think it's somewhat risky from a product pov to allow any dataset names to be editable by default (and perhaps confusing), but if an org wants to opt in this makes a lot a lot of sense.
> 
> Step to add a feature flag:
> 
> 1. Add application.yaml > featureFlags > editableDatasetNames flag
> 2. Add to FeatureFlags.java
> 3. Add to app.graphql AppConfig type in GraphQL the new editableDatasetName field
> 4. Edit AppConfigResolver to map the featureFlag object from YAML to the GraphQL editableDatasetName field.
> 5. Fetch new field inside of app.graphql on the datahub-web-react side
> 6. Add it to DEFAULT_APP_CONFIG in appConfigContext.tsx
> 7. Finally, inside the EntityHeader.tsx component, add useAppConfig to get the feature flags
> 8. Show or hide the edit capability based on whether the flag is true

Thanks for the review and feedback. I will work on this and get back to you.

**Comment by jayasimhankv** [comment]
@jjoyce0510 - implemented feature flag changes to enable editable dataset name. Please review.

Attached some screenshots for reference:
**Application yaml with editable dataset name flag defaulted to false:**
Pasted Graphic 169

**UI does not allow the name to be edited:**
image

**Enabled flag through docker environment var for testing:**
Pasted Graphic 170

**UI allows name to be edited:**
image

**Updated name:**
image

**Comment by jjoyce0510** [comment]
Thanks - taking another pass here now that the feature flags were added

**Comment by rtekal** [comment]
@jjoyce0510 Do you need any more information.  Will you be able to give your comments by end of today please.

**Review by jjoyce0510 — APPROVED** [review]
This looks great. Thanks for the contribution!

**Comment by jjoyce0510** [comment]
Looks only like minor things -- yarn lint appears to be failing.

To comply with linting:

JAVA

```./gradlew datahub-graphql-core:spotlessApply```

TYPESCRIPT

```cd datahub-web-react && yarn prettier --write src && yarn run lint```, then fix issues. If you have "prettier" issues

**Comment by jayasimhankv** [comment]
Thanks for your review. Yes, will address the listing and spotless issues and commit as soon as I am able.

**Comment by jayasimhankv** [comment]
> Looks only like minor things -- yarn lint appears to be failing.
> 
> To comply with linting:
> 
> JAVA
> 
> `./gradlew datahub-graphql-core:spotlessApply`
> 
> TYPESCRIPT
> 
> `cd datahub-web-react && yarn prettier --write src && yarn run lint`, then fix issues. If you have "prettier" issues

Committed files after yarn lint fixes. There weren't any spotless fixes.

**Comment by rtekal** [comment]
@jjoyce0510 Thank you very much for your review.  Hope it can get merged and be part of a release soon.
I can try to repay you... please tag me if I can be of any help.

**Pull request merged**